src/harness/reference_models/dpa/dpa_mgr.py

```diff
@@ -432,6 +451,71 @@ def CheckInterference(self, sas_uut_active_grants, margin_db,
 
     return max_diff_interf <= margin_db
 
+  def __PrintStatistics(self, results, margin_db, dpa_name, channel):
+    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+    filename='%s DPA=%s channel=%s.csv' % (timestamp, self.name, channel)
+    logging.info('Saving stats for DPA %s, channel %s to file: %s', dpa_name, channel, filename)
+    with open(filename, 'w') as f:
+      # CSV header.
+      f.write('Latitude,Longitude,Azimuth (degrees),A_DPA (dBm),A_DPA_ref (dBm),A_DPA - A_DPA_ref,A_DPA + 144 dBm\n')
+      for result, point in zip(results, self.protected_points):
+        latitude = point.latitude
+        longitude = point.longitude
+        for k, azimuth in enumerate(result.azimuth_array):
+          A_DPA = result.A_DPA[k]
+          if isinstance(result.A_DPA_ref, float):  # Happens when there are no peer SASes.
+            A_DPA_ref = result.A_DPA_ref
+          else:
+            A_DPA_ref = result.A_DPA_ref[k]
+          line = ','.join('%3.10f' % val for val in [
+              latitude, longitude, azimuth, A_DPA, A_DPA_ref, A_DPA -
+              A_DPA_ref, A_DPA + 144
+          ])
+          f.write(line + '\n')
+
+    differences = np.zeros([len(self.protected_points), len(results[0].A_DPA)])
+    for k, (result, point) in enumerate(zip(results, self.protected_points)):
+      difference = result.A_DPA - result.A_DPA_ref
+      differences[k, :] = difference
+    logging.info('--- Difference statistics ---')
+    logging.info('Min difference: %s', np.min(differences))
+    logging.info('Max difference: %s', np.max(differences))
+    for percentile in [50, 90, 99, 99.9, 99.99, 99.999, 99.9999]:
+      logging.info('%f percent of differences are <= %f', percentile,
+                   np.percentile(differences, percentile))
+    logging.info('--- End statistics ---')
+
+  def __PrintKeepLists(self, keep_list_th_other_sas, keep_list_th_managing_sas,
+                       keep_list_uut_managing_sas, dpa_name, channel):
+
+    def WriteKeepList(filename, keep_list):
+      logging.info('Writing keep list to file: %s', filename)
+      fields = [
+          'latitude', 'longitude', 'height_agl', 'indoor_deployment',
+          'cbsd_category', 'antenna_azimuth', 'antenna_gain',
+          'antenna_beamwidth', 'max_eirp', 'low_frequency', 'high_frequency',
+          'is_managed_grant'
+      ]
+      with open(filename, 'w') as f:
+        f.write(','.join(fields) + '\n')
+        for cbsd_grant_info in keep_list:
+          cbsd_grant_dict = cbsd_grant_info._asdict()
+          f.write(','.join([str(cbsd_grant_dict[key]) for key in fields]) + '\n')
```

Comment:
If you want you can get as following:
```
for cbsd_grant_info  in ...:
  f.write(','.join(str(getattr(cbsd_grant_info, key)) for key in fields) + '\n')
```
to avoid the conversion to dict and temp creation of a list